aws-crt-kotlin/native/src/aws/sdk/kotlin/crt/Assertions.kt

```diff
@@ -0,0 +1,25 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package aws.sdk.kotlin.crt
+
+import kotlinx.cinterop.ExperimentalForeignApi
+import libcrt.AWS_OP_SUCCESS
+
+/**
+ * Assert that this [Int] return value is `AWS_OP_SUCCESS`. If not, throw [CrtRuntimeException] with the given evaluated
+ * [lazyMessage].
+ */
+@OptIn(ExperimentalForeignApi::class)
+internal inline fun Int.awsAssertOpSuccess(lazyMessage: () -> String) {
+    if (this != AWS_OP_SUCCESS) {
+        throw CrtRuntimeException(lazyMessage())
+    }
+}
+
+/**
+ * Assert that this [T] is non-null. If it is, throw [CrtRuntimeException] with the given evaluated [lazyMessage].
+ */
+internal inline fun <T> T?.awsAssertNotNull(lazyMessage: () -> String): T = this ?: throw CrtRuntimeException(lazyMessage())
```

Comment:
**Suggestion**: How is this going to be used? I don't see any usages right now but, as written, it requires capturing the returned reference to be useful.

```kotlin
val foo = …
val nonNullFoo = foo.awsAssertNotNull { "Whoops!" }
bar(nonNullFoo)
// Can't call `bar(foo)` because the compiler thinks `foo` could be null
```

This could be written using contracts:

```kotlin
@ExperimentalContracts
internal inline fun <T> T?.awsAssertNotNull(lazyMessage: () -> String) {
    contract {
        returns() implies (this != null)
    }
    if (this == null) throw CrtRuntimeException(lazyMessage())
}
```

Would would allow the compiler to infer the receiver is not null:

```kotlin
val foo = …
foo.awsAssertNotNull { "Whoops!" }
bar(foo) // The compiler can now smart-cast `foo` to non-null
```

Reply:
I think we can even remove `awsAssertNotNull`. I want to say I originally added this to ensure all exceptions out of this library were of the type `CrtRuntimeException`. I don't think this is necessary though or even desirable. We can wrap the exception types (and we do already to my knowledge) in consuming code in smithy-kotlin.